src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,316 @@
+import * as child_process from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import { URI } from 'vscode-uri';
+import { Connection } from 'vscode-languageserver';
+import { v4 as uuidv4 } from 'uuid';
+import { asyncExec } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+import { IContianerEngine } from '../interfaces/extensionSettings';
+
+
+export class ExecutionEnvironment {
+    private connection: Connection;
+    private context: WorkspaceFolderContext;
+    private _container_engine: IContianerEngine;
+    private _container_image: string;
+    private _container_image_id: string;
+
+    constructor(connection: Connection, context: WorkspaceFolderContext) {
+      this.connection = connection;
+      this.context = context;
+    }
+
+    public async initialize(): Promise<void> {
+      try {
+        const settings = await this.context.documentSettings.get(
+          this.context.workspaceFolder.uri
+        );
+        this._container_image = settings.executionEnvironment.image
+        this._container_engine = settings.executionEnvironment.containerEngine
+        if (this._container_engine === 'auto') {
+          for (const ce of ['podman', 'docker']) {
+              try {
+                child_process.execSync(`which ${ce}`, {
+                  encoding: 'utf-8'
+                })
+                this._container_engine = <IContianerEngine>ce;
+                this.connection.console.log(`Container engine set to: '${ce}'`)
+                break
+              } catch (error) {
+                this.connection.console.error(`Container engine '${ce}' not found`)
+              }
+            }
+        } else {
+          try {
+            child_process.execSync(`which ${this._container_engine}`, {
+              encoding: 'utf-8'
+            })
+          } catch (error) {
+            this.connection.window.showErrorMessage(`Container engine '${this._container_engine}' not found. Failed with error '${error}'`)
+            return
+          }
+        }
+        if (!['podman', 'docker'].includes(this._container_engine)) {
+          this.connection.window.showInformationMessage(`Valid contianer engine not found ${this._container_engine}.`)
+          return
+        }
+        const imagePuller = new ImagePuller(
+          this.connection,
+          this._container_engine,
+          this._container_image,
+          settings.executionEnvironment.pullPolicy
+        )
+        const setupDone = await imagePuller.setupImage()
+        if (!setupDone) {
+          this.connection.window.showWarningMessage(`Execution environment image ${this._container_image} setup failed.`)
+          return
+        }
+        this.fetchPluginDocs()
+      } catch (error) {
```

Comment:
Same here: using try/except on huge blocks of code is dangerous and it makes it hard to reason about the control flow.

Reply:
The except block is a catch-all errors scenario to bail out any unknown error that might occur instead of stack-trace. I don't better alternative for now.

Reply:
Having a catch-all is a code smell by itself. Beside shadowing unknown problems that makes it hard to debug the issue, it makes control flow rather unpredictable. For example, this code sometimes sets the instance attributes in between various invocations. Depending on which invocation fails, only a part of those attributes will be set, meaning that the object may end up in a weird state and nobody would notice.
Instead of using a catch-all, we should propagate all the exceptions up the stack. It is only acceptable to catch them in the outer-most layer, when necessary. But in general, the unexpected cases should fail loudly.

Reply:
The initialise method can be considered as top of the stack for exeuctionEnvironment service. Any unknown failures can be handled gracefully here. If it is not done the langauge-server might end up crashing and displaying  the crash error in the extension window which will be a bad user experience.

Reply:
In the scenario you describe, the service is the top of the stack. And this is where you could have a try/catch thing that wraps exactly one line. Currently, this is on the wrong layer.

Reply:
I don't think so it is at wrong layer. Besides I am not of the opinion that try/catch should warp only single line always. You can refer similar try/catch pattern for other services in the same code base.